Mathlib/NumberTheory/Padics/Harmonic.lean

```diff
@@ -0,0 +1,139 @@
+/-
+Copyright (c) 2023 Koundinya Vajjha. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Koundinya Vajjha
+-/
+
+import Mathlib.NumberTheory.Padics.PadicIntegers
+import Mathlib.Data.Int.Log
+/-!
+
+The nth Harmonic number is not an integer. We formalize the proof using
+2-adic valuations. This proof is due to Kürschák.
+
+Reference:
+https://kconrad.math.uconn.edu/blurbs/gradnumthy/padicharmonicsum.pdf
+
+-/
+
+open BigOperators
+
+/-- The nth-harmonic number defined as a finset sum of consecutive reciprocals. -/
+def harmonic : ℕ → ℚ := fun n => ∑ i in Finset.range n, 1 / (i + 1)
+
+lemma harmonic_pos {n : ℕ} (Hn : n ≠ 0) : 0 < harmonic n :=
+  Finset.sum_pos (fun _ _ => div_pos zero_lt_one (by norm_cast; linarith))
+  (by rwa [Finset.nonempty_range_iff])
+
+lemma harmonic_singleton {n c : ℕ} (hc : c ∈ Finset.range n) :
+    harmonic n = 1 / (c + 1) + ∑ x in Finset.range n \ {c}, 1 / ((x : ℚ) + 1) := by
+  dsimp [harmonic]
+  rwa [add_comm, Finset.sum_eq_sum_diff_singleton_add (i := c)]
+
+lemma finset_range_sdiff_singleton_nonempty {c n : ℕ} (hn : 2 ≤ n) :
+    Finset.Nonempty (Finset.range n \ {c}) := by
+  rw [Finset.sdiff_nonempty, Finset.subset_singleton_iff, Finset.range_eq_empty_iff, not_or]
+  refine' ⟨by linarith, fun Hnot => _⟩
+  suffices n = 1 by linarith
+  rw [← Finset.card_range n, ← Finset.card_singleton c, Hnot]
+
+lemma harmonic_singleton_ne_zero {c n : ℕ} (hn : 2 ≤ n) :
+    ∑ x in Finset.range n \ {c}, 1 / (x + 1 : ℚ) ≠ 0 := by
+  refine' ne_of_gt <| Finset.sum_pos (fun i _ => div_pos zero_lt_one _)
+    (finset_range_sdiff_singleton_nonempty hn)
+  norm_cast; simp only [add_pos_iff, or_true]
+
+lemma padicValRat_two_pow_div (r : ℕ) : padicValRat 2 (1 / 2 ^ r) = -r := by
```

Comment:
This is already done in 38884190410085716619b1cf03d126b033546b90. See `padicValRat.self_pow_div`. I use this lemma twice in two proofs below so I factored it out as a local lemma. I can't seem to use `self_pow_div` directly because of an implicit coercion in the proof below.